netbox/generate_basic_tests.go

```diff
@@ -0,0 +1,383 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//+build ignore
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/format"
+	"log"
+	"os"
+	"text/template"
+	"time"
+)
+
+func main() {
+	typeName := flag.String("type-name", "Example", "Name of the type to use (e.g. TenantGroup).")
+	serviceName := flag.String("service-name", "ExampleService", "Name of the service to create (e.g. TenantGroupsService).")
+	endpoint := flag.String("endpoint", "tenancy", "Name of the endpoint (e.g. dcim, ipam, tenancy).")
+	service := flag.String("service", "example", "Name of the service below endpoint (e.g. tenant-groups).")
+	clientEndpoint := flag.String("client-endpoint", "Tenancy", "Name of the client endpoint (e.g. DCIM, IPAM, Tenancy).")
+	clientService := flag.String("client-service", "TenantGroups", "Name of the client service (e.g. TenantGroups, Tenants).")
+	withoutListOpts := flag.Bool("without-list-opts", false, "Disable list options for this endpoint.")
+
+	flag.Parse()
+
+	b := &bytes.Buffer{}
+	functionsTemplate.Execute(b, struct {
+		Timestamp      time.Time
+		TypeName       string
+		ServiceName    string
+		Endpoint       string
+		Service        string
+		ClientEndpoint string
+		ClientService  string
+		ListOpts       bool
+		JSONTag        func(string) string
+	}{
+		Timestamp:      time.Now(),
+		TypeName:       *typeName,
+		ServiceName:    *serviceName,
+		Endpoint:       *endpoint,
+		Service:        *service,
+		ClientEndpoint: *clientEndpoint,
+		ClientService:  *clientService,
+		ListOpts:       !*withoutListOpts,
+		JSONTag:        func(name string) string { return "`json:\"" + name + "\"`" },
+	})
+
+	// go fmt
+	res, err := format.Source(b.Bytes())
+	if err != nil {
+		b.WriteTo(os.Stderr)
+		log.Fatal(err)
+	}
+
+	f, err := os.Create(fmt.Sprintf("%s_%s_basic_test.go", *endpoint, *service))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	f.Write(res)
+}
+
+var functionsTemplate = template.Must(template.New("").Parse(`// Copyright 2017 The go-netbox Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// GENERATED BY THE COMMAND ABOVE; DO NOT EDIT
```

Comment:
Let's make this match the new generated code detection header: https://github.com/golang/go/issues/13560#issuecomment-288457920.